### ISSUE TITLE
Hygiene bug in disintegration

### ISSUE THREAD

**Issue by wrengr** [hakaru-dev/hakaru#12]
In [`Tests.Disintegrate`](https://github.com/hakaru-dev/hakaru/blob/master/haskell/Tests/Disintegrate.hs#L125), look at the first solution returned by `testPerform1b` and `testPerform1c` (also note that `testPerform1a` doesn't suffer the bug). The first statement emitted for that solution just throws its variable away, and what should be the use sites of that variable now point to a free variable instead.

In particular, this bug has been tracked down to the `performWhnf` helper function ([`Language.Hakaru.Disintegrate`:372–375](https://github.com/hakaru-dev/hakaru/blob/master/haskell/Language/Hakaru/Disintegrate.hs#L372)). Namely, that function's call to `emitMBind` is invalid because the expression being emitted contains heap-bound variables. For these tests in particular, that call to `emitMBind` _should_ actually be acceptable— the heap-bound variables in question are `SLet`-bound to a variable which has already been emitted, thus we should just perform the variable renaming before emitting (or, have already performed the renaming whenever we emitted those variables).

**Comment by ccshan** [comment]
For understanding and debugging, it seems useful to document (in the form of a grammar) what terms (and in what scope) each function might take as input and return as output, no?

**Comment by JacquesCarette** [comment]
Is this still a live issue?

**Comment by ccshan** [comment]
The problem in `testPerform1b` and `testPerform1c` that @wrengr describes is still there (I just checked), though I'm not sure how to make the problem visible as a bug in the `disintegrate` command. For example, the following concrete Hakaru seems to contain `norm1b`/`norm1c` and seems to `disintegrate` fine:
```
zu <~ x <~ normal(3,2)
      if x < 0:
        return (-x, ())
      else:
        return (x, ())
match zu:
 (z,u):
  y <~ normal (z*z,1)
  return (y,())
```
Maybe @pravnar can decide whether to close or re-open the issue?

**Comment by pravnar** [comment]
We fix this by going under "Neutral" terms (c420a10). 
The results of `testPerform1b` and `testPerform1c` don't show hygiene issues now (f05b829).
